### PR TITLE
Basic Savestate

### PR DESCRIPTION
**Pull request by Piiman**
#4 

Added a Savestates button in the main menu and its corresponding screen (images attached)

Main menu
2024-04-09_14-34-50 440_bot

Savestates menu
2024-04-10_02-56-18 165_bot

The arrows cycle throgh the save slots, the load button executes the `emulation_lstate` function, save button the `emulation_sstate` function

The savestate file is created inside the directory `vb_savestates/{rom name}/` it generates the directory in the same address where the rom is

As of now for each rom you'll have access to a whopping 10 savestates

This was tested with Jack Bros and Mario's tennis

### CODE REVIEW

**Comment by skyfloogle** [comment]
This is a good start! There's a few things that currently aren't recorded in the savestates such as the sound state and some static variables in the interrupt checkers, and I don't want to put this out to the world before that's sorted out, but the UI side looks good.

**Comment by Piiman** [comment]
Thanks for checking the PR!

Yes, I noticed that when you load a savestate the sound gets lost until another track gets loaded, and for the ui I also wanted it to display the time and date of the save below the text `State X`, but I haven't had much free time this couple of days, might look into it on the week end

Also, regarding the directory in which the savestates are stored, I was wondering if it would be better to save them inside a folder at the root of the sd instead of where the rom is located, this because the function `get_savestate_path` turned out more complicated than I thought it would be, what do you think?

File: source/common/vb_gui.c [hunk 1]
```diff
@@ -343,14 +347,38 @@ int emulation_reset(void) {
     return D_EXIT;
 }
 
-int emulation_sstate(void) {
+char * get_savestate_path(int state) {
+    char * sspath;
+    sspath = (char *) malloc(131 * sizeof(char));
```
Comment (eku) [inline source/common/vb_gui.c:352]: Where does the 131 come from? Is it ensured that the following `strcat` do not write beyond the end?
What speaks against the use of a single `snprintf` that assembles the complete path

File: source/common/vb_gui.c [hunk 2]
```diff
@@ -418,14 +446,12 @@ int emulation_sstate(void) {
...
-int emulation_lstate(void) {
+int emulation_lstate(int state) {
     int i;
     int ret;
     int id,ver,crc;
-    char sspath[131];
     FILE* state_file;
-
-    sprintf(sspath, "%s.rds", tVBOpt.ROM_NAME);
+    char* sspath = get_savestate_path(state);
```
Comment (eku) [inline source/common/vb_gui.c:454]: When will the memory be released?

File: source/common/vb_gui.c [hunk 3]
```diff
@@ -343,14 +347,38 @@ int emulation_reset(void) {
...
+
+int emulation_sstate(int state) {
     int i;
     int highbyte;
     int lowbyte;
-    char sspath[131];
     FILE* state_file;
-
-    sprintf(sspath, "%s.rds", tVBOpt.ROM_NAME);
+    char* sspath = get_savestate_path(state);
```
Comment (eku) [inline source/common/vb_gui.c:381]: When will the memory be released?

**Review by eku — CHANGES_REQUESTED** [review]

File: source/common/vb_gui.c [hunk 4]
```diff
@@ -343,14 +347,38 @@ int emulation_reset(void) {
...
+char * get_savestate_path(int state) {
+    char * sspath;
+    sspath = (char *) malloc(131 * sizeof(char));
+    char file_name[8];
+    char * last_slash = strrchr(tVBOpt.ROM_PATH, '/');
+    strncpy(sspath, tVBOpt.ROM_PATH, last_slash - tVBOpt.ROM_PATH);
+    sspath[last_slash - tVBOpt.ROM_PATH] = '\0';
+    strcat(sspath, "/vb_savestates");
+    if(stat(sspath, &st) == -1)
+        mkdir(sspath, 0777);
```
Comment (eku) [inline source/common/vb_gui.c:359]: Return value of `mkdir` not checked!

File: source/common/vb_gui.c [hunk 5]
```diff
@@ -343,14 +347,38 @@ int emulation_reset(void) {
...
+    sprintf(file_name, "/%d.rds", state);
+    strcat(sspath, file_name);
+
+    return sspath;
+}
+
+int emulation_rmstate(int state) {
+    char* sspath = get_savestate_path(state);
```
Comment (eku) [inline source/common/vb_gui.c:371]: When will the memory be released?

**Review by eku — CHANGES_REQUESTED** [review]

**Comment by skyfloogle** [comment]
I'd generally go with the precedent set by other emulators.
That said, there's a bunch of things it would need to save, considering other frequently requested features. Global settings, per-game settings, savestates, output from the 3dslink clone, potentially debug dumps. Maybe it's best to make a directory for Red Viper to put its stuff in, like sdmc:/red-viper?

**Comment by Piiman** [comment]
sounds good, I understand that there would be a subdirectory or file for these features (`sdmc:/red-viper/savestates/`, `sdmc:/red-viper/settings.ini`, `sdmc:/red-viper/debug/`, etc), I'll implement it just for savestates for now, this would adding the path to the current settings file, creating the base directory on startup and creating the savestates directory (don't know the timing yet tho), but overall I think this would be a more standard practice and would make implementing new features easier in the future.

File: source/common/vb_gui.c (hunk 3 above)
Comment (Piiman) [reply source/common/vb_gui.c:381]: addressed in [1561bf](https://github.com/Piiman/red-viper/commit/18561bffe44680dbee11116bd3e8412ae1ec0f1a)

File: source/common/vb_gui.c (hunk 2 above)
Comment (Piiman) [reply source/common/vb_gui.c:454]: addressed in [1561bf](https://github.com/Piiman/red-viper/commit/18561bffe44680dbee11116bd3e8412ae1ec0f1a)

File: source/common/vb_gui.c (hunk 5 above)
Comment (Piiman) [reply source/common/vb_gui.c:371]: addressed in [9f26379](https://github.com/skyfloogle/red-viper/pull/59/commits/9f26379e5c1fc95c22ec9355840a60331aff69de)

File: include/v810_mem.h [hunk 6]
```diff
@@ -5,27 +5,31 @@
...
+        HWORD tFrameBuffer;   //not publicly visible
+        HWORD tFrame;
+        HWORD VER;
+        HWORD SPT[4];
+        HWORD GPLT[4];
+        HWORD JPLT[4];
+        HWORD BKCOL;
+    };
+
+    HWORD V810_VIPREGARR[20];
```
Comment (skyfloogle) [inline include/v810_mem.h:32]: Is this needed for anything? If it's intended to treat the whole struct as one array, it's incorrect, as while there are 20 variables, some of them are arrays.

Comment (Piiman) [reply include/v810_mem.h]: No, it's not needed, it was intended for that, but I wasn't able to make it work and forgot to rollback this specific part of it, shoul be reverted in [50c4918](https://github.com/skyfloogle/red-viper/pull/59/commits/50c4918ab30979da68f07b02d38827556a11cf80)

**Comment by eku** [comment]
Please check for return codes of library calls and act on errors instead of crashing the emulator.

**Comment by skyfloogle** [comment]
I think this is in a mergeable state now.

**Comment by eku** [comment]
Too bad that return values are not checked and the emulator crashes. There was no lack of hints for improving the code.

**Comment by skyfloogle** [comment]
I did make some improvements. I checked the longest possible file path and allocated with that in mind, and I changed the savestate format so that it's checked continuously for validity while loading and not applied until it's confirmed to be valid. Any failure during savestate code will simply show an error screen, and return to the emulator cleanly. Everything specific that you pointed out has been addressed, except for malloc calls.
As far as malloc goes, the emulator is pretty careful about allocation. Most allocation happens at boot time. The only allocations after that are in menu-related code, and they're small allocations that are freed as necessary. If malloc fails, I honestly want the emulator to crash, as that indicates a programming error. Of course, ideally everything would be statically allocated, but I haven't investigated the implications of that.
If you feel that I missed something important, you're welcome to send in a pull request.

**Comment by Magnusretrotech** [comment]
Thank you so much for this!

**Pull request merged**